src/history/hs.c

```diff
@@ -561,39 +531,30 @@ __wt_hs_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_RECONCILE *r, WT_M
          */
         WT_ASSERT(session, modifies.size > 0);
         __wt_modify_vector_pop(&modifies, &upd);
-
-        /* If we popped a modify then it should be flagged as in the history store. */
-        if (upd->type == WT_UPDATE_MODIFY) {
-            WT_ASSERT(session, F_ISSET(upd, WT_UPDATE_HS));
-            __wt_modify_vector_peek(&modifies, &prev_upd);
-            /*
-             * Retrieve the full value of the modify from the history store. This avoid us having to
-             * iterate the full update list associated with the modify and recalculating the reverse
-             * deltas.
-             */
-            WT_ERR(__hs_get_value(
-              cursor, btree, upd, prev_upd->start_ts, prev_upd->txnid, key, full_value));
-        } else {
-            /* The key didn't exist back then, which is globally visible. */
-            WT_ASSERT(session, upd->type == WT_UPDATE_STANDARD || upd->type == WT_UPDATE_TOMBSTONE);
-        }
-
         /* Skip TOMBSTONE at the end of the update chain. */
         if (upd->type == WT_UPDATE_TOMBSTONE) {
             if (modifies.size > 0) {
                 __wt_modify_vector_pop(&modifies, &upd);
-                WT_ASSERT(session, upd->type == WT_UPDATE_STANDARD);
             } else
                 continue;
         }
+
+        /* The key didn't exist back then, which is globally visible. */
+        WT_ASSERT(session, upd->type == WT_UPDATE_STANDARD);
         full_value->data = upd->data;
         full_value->size = upd->size;
 
         squashed = false;
 
-        /* Flush the updates on stack. */
-        for (; modifies.size > 0; tmp = full_value, full_value = prev_full_value,
-                                  prev_full_value = tmp, upd = prev_upd) {
+        /*
+         * Flush the updates on stack. Stopping once we run out or we reach the onpage upd start
+         * time pair, we can squash modifies with the same start time pair as the onpage upd away.
+         */
+        for (; modifies.size > 0 &&
+             !(upd->txnid == list->onpage_upd->txnid &&
+                 upd->start_ts == list->onpage_upd->start_ts);
```

Comment:
may be (upd->txnid != list->onpage_upd->txnid ||
                 upd->start_ts != list->onpage_upd->start_ts)